File: backend/app/services/automation_service.py
from __future__ import annotations

import json
import logging
from datetime import UTC, datetime
from decimal import Decimal
from typing import Any

from app.db.supabase import DatabaseConfigError, DatabaseQueryError
from app.executors.email_executor import EmailExecutor
from app.executors.github_executor import GitHubExecutor
from app.executors.mock_cloud_executor import MockCloudExecutor
from app.executors.slack_executor import SlackExecutor
from app.models.enums import ActionStatus
from app.rules.engine import SAMPLE_RULE, rule_engine
from app.schemas.actions import ActionCreate
from app.schemas.approvals import ApprovalCreate
from app.services.database import database_service

logger = logging.getLogger(__name__)


SAFE_MODES = {"simulation", "dry_run", "safe_execute"}


class AutomationService:
    """Coordinates rule decisions, approvals, executors, and action logs."""

    def create_rule(self, payload) -> dict[str, Any]:
        return database_service.create_rule(payload)

    def list_rules(self, limit: int = 100) -> dict[str, Any]:
        try:
            return {"rules": database_service.list_rules(limit=limit), "sample_rule": SAMPLE_RULE}
        except (DatabaseConfigError, DatabaseQueryError) as exc:
            return {"rules": [], "sample_rule": SAMPLE_RULE, "errors": [str(exc)]}

    async def run_action(
        self,
        *,
        finding_id: str,
        action_type: str | None = None,
        execution_mode: str = "dry_run",
        slack_webhook_url: str | None = None,
        recipient_email: str | None = None,
    ) -> dict[str, Any]:
        if execution_mode not in SAFE_MODES:
            raise ValueError(f"Unsupported execution mode: {execution_mode}")

        finding = database_service.get_ai_finding(finding_id)
        if not finding:
            raise ValueError("Finding not found")

        rules = database_service.list_rules()
        decisions = rule_engine.evaluate_findings([finding], rules)
        selected = decisions[0] if decisions else {
            "execution_mode": "approval_required",
            "action_type": action_type or "manual_review",
            "rule": None,
        }

        resolved_action_type = action_type or selected["action_type"]
        resolved_mode = selected["execution_mode"]

        if resolved_mode == "approval_required":
            action = self._log_action(
                finding=finding,
                action_type=resolved_action_type,
                status=ActionStatus.PENDING,
                execution_mode="approval_required",
                result={"message": "Approval required before execution."},
            )
            approval = database_service.create_approval(
                ApprovalCreate(finding_id=finding_id, action_id=action["id"], status="pending")
            )
            return {"mode": "approval_required", "approval": approval, "action": action, "decision": selected}

        if resolved_mode == "alert_only":
            result = await self._execute(
                action_type="slack_alert" if resolved_action_type == "manual_review" else resolved_action_type,
                finding=finding,
                execution_mode="dry_run" if execution_mode == "simulation" else execution_mode,
                slack_webhook_url=slack_webhook_url,
                recipient_email=recipient_email,
            )
            action = self._log_action(
                finding=finding,
                action_type=resolved_action_type,
                status=ActionStatus.COMPLETED,
                execution_mode="alert_only",
                result=result,
            )
            return {"mode": "alert_only", "action": action, "execution": result, "decision": selected}

        result = await self._execute(
            action_type=resolved_action_type,
            finding=finding,
            execution_mode=execution_mode,
            slack_webhook_url=slack_webhook_url,
            recipient_email=recipient_email,
        )
        action = self._log_action(
            finding=finding,
            action_type=resolved_action_type,
            status=ActionStatus.COMPLETED,
            execution_mode=execution_mode,
            result=result,
        )
        return {"mode": "auto_execute", "action": action, "execution": result, "decision": selected}

    def list_action_logs(self, limit: int = 100) -> dict[str, Any]:
        try:
            return {"actions": database_service.list_actions(limit=limit)}
        except (DatabaseConfigError, DatabaseQueryError) as exc:
            return {"actions": [], "errors": [str(exc)]}

    async def _execute(
        self,
        *,
        action_type: str,
        finding: dict[str, Any],
        execution_mode: str,
        slack_webhook_url: str | None,
        recipient_email: str | None,
    ) -> dict[str, Any]:
        if action_type in {"slack_alert", "alert"}:
            return await SlackExecutor().execute(
                finding=finding,
                webhook_url=slack_webhook_url,
                mode=execution_mode,
            )
        if action_type in {"github_remove_seat", "remove_github_seat"}:
            return GitHubExecutor().execute(finding=finding, mode=execution_mode)
        if action_type in {"cloud_shutdown", "cloud_resize", "mock_cloud_action"}:
            return MockCloudExecutor().execute(finding=finding, mode=execution_mode)
        if action_type in {"email", "email_owner"}:
            return EmailExecutor().execute(
                finding=finding,
                recipient=recipient_email,
                mode=execution_mode,
            )
        return EmailExecutor().execute(finding=finding, recipient=recipient_email, mode="simulation")

    @staticmethod
    def _log_action(
        *,
        finding: dict[str, Any],
        action_type: str,
        status: ActionStatus,
        execution_mode: str,
        result: dict[str, Any],
    ) -> dict[str, Any]:
        return database_service.create_action(
            ActionCreate(
                finding_id=finding["id"],
                action_type=action_type,
                status=status,
                executed_at=datetime.now(UTC),
                savings=Decimal(str(finding.get("estimated_savings") or "0")),
                execution_log=result.get("execution_log") or result.get("message") or json.dumps(result),
                execution_result=json.dumps(result),
                execution_mode=execution_mode,
                rollback_status="not_required",
            )
        )


automation_service = AutomationService()

File: backend-java/src/main/java/com/killswitch/service/AutomationService.java
package com.killswitch.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.killswitch.executor.GitHubExecutor;
import com.killswitch.executor.SlackExecutor;
import com.killswitch.executor.StripeExecutor;
import com.killswitch.rules.RuleEngine;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AutomationService {
    private final DatabaseService database;
    private final RuleEngine ruleEngine;
    private final SlackExecutor slackExecutor;
    private final GitHubExecutor gitHubExecutor;
    private final StripeExecutor stripeExecutor;
    private final ObjectMapper objectMapper;

    public Map<String, Object> createRule(Map<String, Object> payload) {
        return database.createRule(payload);
    }

    public Map<String, Object> listRules(int limit) {
        try {
            return Map.of("rules", database.listRules(limit), "sample_rule", sampleRule());
        } catch (Exception e) {
            return Map.of("rules", List.of(), "sample_rule", sampleRule(), "errors", List.of(e.getMessage()));
        }
    }

    public Map<String, Object> updateRule(String id, Map<String, Object> payload) {
        return database.updateRule(id, payload);
    }

    public void deleteRule(String id) {
        database.deleteRule(id);
    }

    public Map<String, Object> runAction(String findingId, String actionType, String executionMode,
        String slackWebhookUrl, String recipientEmail) {
        if (!List.of("simulation", "dry_run", "safe_execute").contains(executionMode)) {
            throw new IllegalArgumentException("Unsupported execution mode: " + executionMode);
        }
        Map<String, Object> finding = database.getFinding(findingId);
        if (finding == null) {
            throw new IllegalArgumentException("Finding not found");
        }

        List<RuleEngine.Decision> decisions = ruleEngine.evaluate(List.of(finding), database.listRules(100));
        RuleEngine.Decision decision = decisions.isEmpty()
            ? new RuleEngine.Decision(finding, null, actionType == null ? "manual_review" : actionType, "approval_required")
            : decisions.get(0);

        String resolvedAction = actionType == null || actionType.isBlank() ? decision.actionType() : actionType;
        String resolvedMode = decision.executionMode();
        if ("approval_required".equals(resolvedMode)) {
            Map<String, Object> action = logAction(finding, resolvedAction, "pending", "approval_required",
                Map.of("message", "Approval required before execution."));
            Map<String, Object> approval = database.createApproval(Map.of(
                "finding_id", findingId,
                "action_id", action.get("id"),
                "status", "pending"
            ));
            return Map.of("mode", "approval_required", "approval", approval, "action", action, "decision", decision.toMap());
        }

        Map<String, Object> result = execute(resolvedAction, finding, executionMode, slackWebhookUrl);
        Map<String, Object> action = logAction(finding, resolvedAction, "completed", executionMode, result);
        return Map.of("mode", "auto_execute", "action", action, "execution", result, "decision", decision.toMap());
    }

    public Map<String, Object> listActionLogs(int limit) {
        try {
            return Map.of("actions", database.listActions(limit));
        } catch (Exception e) {
            return Map.of("actions", List.of(), "errors", List.of(e.getMessage()));
        }
    }

    public Map<String, Object> getAction(String id) {
        return database.getAction(id);
    }

    public Map<String, Object> approve(String approvalId, String approvedBy) {
        return setApprovalStatus(approvalId, "approved", approvedBy);
    }

    public Map<String, Object> reject(String approvalId, String approvedBy) {
        return setApprovalStatus(approvalId, "rejected", approvedBy);
    }

    public Map<String, Object> listPendingApprovals(int limit) {
        try {
            return Map.of("approvals", database.listPendingApprovals(limit));
        } catch (Exception e) {
            return Map.of("approvals", List.of(), "errors", List.of(e.getMessage()));
        }
    }

    private Map<String, Object> setApprovalStatus(String approvalId, String status, String approvedBy) {
        Map<String, Object> existing = database.getApproval(approvalId);
        if (existing == null) {
            throw new IllegalArgumentException("Approval request not found");
        }
        Map<String, Object> update = new LinkedHashMap<>();
        update.put("status", status);
        update.put("approved_by", approvedBy);
        update.put("approved_at", OffsetDateTime.now());
        Map<String, Object> updated = database.updateApproval(approvalId, update);
        Object actionId = existing.get("action_id");
        if (actionId != null) {
            database.updateAction(String.valueOf(actionId), Map.of(
                "status", "rejected".equals(status) ? "skipped" : "pending",
                "execution_log", "rejected".equals(status)
                    ? "Approval rejected; action will not execute."
                    : "Approval granted; action is ready for safe execution."
            ));
        }
        return updated == null ? existing : updated;
    }

    private Map<String, Object> execute(String actionType, Map<String, Object> finding, String mode, String slackWebhookUrl) {
        if (List.of("slack_alert", "alert", "manual_review").contains(actionType)) {
            return slackExecutor.execute(finding, slackWebhookUrl, mode);
        }
        if (List.of("github_remove_seat", "remove_github_seat", "remove_seat").contains(actionType)) {
            return gitHubExecutor.execute(finding, mode);
        }
        if (List.of("stripe_cancel_subscription", "cancel").contains(actionType)) {
            return stripeExecutor.execute(finding, mode);
        }
        return Map.of("executed", false, "mode", "simulation", "execution_log", "No executor exists for " + actionType);
    }

    private Map<String, Object> logAction(Map<String, Object> finding, String actionType, String status,
        String executionMode, Map<String, Object> result) {
        try {
            return database.createAction(Map.of(
                "finding_id", finding.get("id"),
                "action_type", actionType,
                "status", status,
                "executed_at", OffsetDateTime.now(),
                "savings", new BigDecimal(String.valueOf(finding.getOrDefault("estimated_savings", "0"))),
                "execution_mode", executionMode,
                "execution_log", String.valueOf(result.getOrDefault("execution_log", result.getOrDefault("message", ""))),
                "execution_result", objectMapper.writeValueAsString(result),
                "rollback_status", "not_required"
            ));
        } catch (Exception e) {
            throw new RuntimeException("Unable to log action: " + e.getMessage(), e);
        }
    }

    private Map<String, Object> sampleRule() {
        return Map.of(
            "condition_type", "estimated_savings",
            "threshold", "100",
            "action", "manual_review",
            "approval_required", true,
            "enabled", true
        );
    }
}
